File: roulette.rb
#!/usr/bin/env ruby
# roulette.rb
# encoding: UTF-8

require 'json'
require 'fileutils'
require 'timeout'

COLORS = {
  reset: "\e[0m",
  green: "\e[92m",
  red: "\e[91m",
  yellow: "\e[93m",
  blue: "\e[94m",
  cyan: "\e[96m",
  bold: "\e[1m"
}

def colorize(text, color)
  "#{COLORS[color]}#{text}#{COLORS[:reset]}"
end

class Roulette
  attr_reader :mode, :players_count, :bullets, :chambers, :delay, :names, :stats, :stats_file

  def initialize(mode, players, bullets, chambers, delay, names)
    @mode = mode
    @players_count = players
    @bullets = bullets
    @chambers = chambers
    @delay = delay
    @names = names || (1..players).map { |i| "Игрок #{i}" }
    @chamber = []
    @position = 0
    @current = 0
    @history = []
    @stats_file = File.join(Dir.home, '.roulette_stats.json')
    load_stats
  end

  def load_stats
    if File.exist?(@stats_file)
      @stats = JSON.parse(File.read(@stats_file))
    else
      @stats = {}
    end
  end

  def save_stats
    File.write(@stats_file, JSON.pretty_generate(@stats))
  end

  def display_stats
    if @stats.empty?
      puts colorize("Статистика пуста.", :yellow)
      return
    end
    @stats.each do |name, data|
      puts colorize("#{name}:", :bold)
      puts "  Игр: #{data['games']}, Побед: #{data['wins']}, Поражений: #{data['losses']}, Очки: #{data['score']}"
    end
  end

  def load_chamber
    @chamber = [0] * @chambers
    positions = (0...@chambers).to_a.shuffle
    @bullets.times { |i| @chamber[positions[i]] = 1 }
    @position = 0
  end

  def spin
    @position = rand(@chambers)
  end

  def fire(player)
    if @chamber[@position] == 1
      @history << { player: player, shot: true }
      return true
    else
      @history << { player: player, shot: false }
      @position = (@position + 1) % @chambers
      return false
    end
  end

  def update_stats(player, win, extra = 0)
    name = @names[player]
    @stats[name] ||= { 'games' => 0, 'wins' => 0, 'losses' => 0, 'score' => 0 }
    @stats[name]['games'] += 1
    if win
      @stats[name]['wins'] += 1
    else
      @stats[name]['losses'] += 1
    end
    @stats[name]['score'] += extra
    save_stats
  end

  def all_players_shot?
    return false if @history.size < @players_count
    last = @history.last(@players_count).map { |h| h[:player] }.uniq
    last.size == @players_count
  end

  def play_round
    if @mode == 'classic'
      load_chamber
      @position = 0
      loop do
        player = @current
        puts colorize("\nХод игрока #{@names[player]}", :cyan)
        print "Нажмите Enter, чтобы выстрелить..."
        STDIN.gets
        sleep @delay if @delay > 0
        shot = fire(player)
        if shot
          puts colorize("💥 БАХ! #{@names[player]} проиграл!", :red)
          update_stats(player, false)
          @players_count.times { |i| update_stats(i, true) if i != player }
          return
        else
          puts colorize("Щелчок...", :yellow)
          @current = (@current + 1) % @players_count
          if all_players_shot?
            puts colorize("Все выжили! Ничья.", :blue)
            @players_count.times { |i| update_stats(i, false) }
            return
          end
        end
      end
    elsif @mode == 'spin'
      load_chamber
      loop do
        player = @current
        puts colorize("\nХод игрока #{@names[player]}", :cyan)
        spin
        print "Нажмите Enter, чтобы выстрелить..."
        STDIN.gets
        sleep @delay if @delay > 0
        shot = fire(player)
        if shot
          puts colorize("💥 БАХ! #{@names[player]} проиграл!", :red)
          update_stats(player, false)
          @players_count.times { |i| update_stats(i, true) if i != player }
          return
        else
          puts colorize("Щелчок...", :yellow)
          @current = (@current + 1) % @players_count
          if all_players_shot?
            puts colorize("Все выжили! Ничья.", :blue)
            @players_count.times { |i| update_stats(i, false) }
            return
          end
        end
      end
    elsif @mode == 'duel'
      load_chamber
      @position = 0
      loop do
        player = @current
        puts colorize("\nДуэль! Ход #{@names[player]}", :cyan)
        print "Нажмите Enter, чтобы выстрелить..."
        STDIN.gets
        sleep @delay if @delay > 0
        shot = fire(player)
        if shot
          puts colorize("💥 #{@names[player]} проиграл дуэль!", :red)
          update_stats(player, false)
          other = 1 - player
          update_stats(other, true)
          return
        else
          puts colorize("Щелчок...", :yellow)
          @current = (@current + 1) % @players_count
        end
      end
    elsif @mode == 'survival'
      load_chamber
      @position = 0
      player = 0
      score = 0
      puts colorize("\nРежим выживания для #{@names[player]}", :blue)
      loop do
        puts colorize("Текущий счёт: #{score}", :yellow)
        print "Выстрелить? (y/n): "
        choice = STDIN.gets.chomp.strip.downcase
        if choice == 'q'
          puts "Выход."
          return
        end
        if choice == 'n'
          puts colorize("Вы остановились с #{score} очками.", :green)
          update_stats(player, true, score)
          return
        end
        if choice == 'y'
          sleep @delay if @delay > 0
          shot = fire(player)
          if shot
            puts colorize("💥 БАХ! Вы проиграли! Очки: #{score}", :red)
            update_stats(player, false, score)
            return
          else
            puts colorize("Щелчок...", :yellow)
            score += 1
            @position = (@position + 1) % @chambers
          end
        end
      end
    end
  end

  def play
    puts colorize("🔫 Добро пожаловать в русскую рулетку!", :bold)
    puts "Режим: #{@mode}, Игроков: #{@players_count}, Патронов: #{@bullets}, Гнёзд: #{@chambers}"
    puts "Задержка: #{@delay * 1000} мс" if @delay > 0
    puts "Имена: #{@names.join(', ')}"
    puts
    play_round
  end
end

def main
  mode = 'classic'
  players = 2
  bullets = 1
  chambers = 6
  delay = 0.2
  names = nil
  show_stats = false

  i = 0
  while i < ARGV.size
    arg = ARGV[i]
    case arg
    when 'classic', 'spin', 'duel', 'survival'
      mode = arg
    when '-p'
      players = ARGV[i+1].to_i
      i += 1
    when '-b'
      bullets = ARGV[i+1].to_i
      i += 1
    when '-c'
      chambers = ARGV[i+1].to_i
      i += 1
    when '-n'
      names = ARGV[i+1].split(',')
      i += 1
    when '-d'
      delay = ARGV[i+1].to_f
      i += 1
    when '-s'
      show_stats = true
    when '-h'
      puts "Usage: ruby roulette.rb [mode] [-p players] [-b bullets] [-c chambers] [-n names] [-d delay] [-s]"
      return
    end
    i += 1
  end
  names ||= (1..players).map { |i| "Игрок #{i}" }
  game = Roulette.new(mode, players, bullets, chambers, delay, names)
  if show_stats
    game.display_stats
    return
  end
  game.play
end

main if __FILE__ == $0

File: roulette.js
// roulette.js
#!/usr/bin/env node
'use strict';

const readline = require('readline');
const fs = require('fs');
const path = require('path');
const os = require('os');

const COLORS = {
    reset: '\x1b[0m',
    green: '\x1b[92m',
    red: '\x1b[91m',
    yellow: '\x1b[93m',
    blue: '\x1b[94m',
    cyan: '\x1b[96m',
    bold: '\x1b[1m'
};

function colorize(text, color) {
    return COLORS[color] + text + COLORS.reset;
}

class Roulette {
    constructor(mode, players, bullets, chambers, delay, names) {
        this.mode = mode;
        this.playersCount = players;
        this.bullets = bullets;
        this.chambers = chambers;
        this.delay = delay;
        this.names = names || Array.from({length: players}, (_,i) => `Игрок ${i+1}`);
        this.chamber = [];
        this.position = 0;
        this.current = 0;
        this.history = [];
        this.statsFile = path.join(os.homedir(), '.roulette_stats.json');
        this.stats = this.loadStats();
    }

    loadStats() {
        try {
            return JSON.parse(fs.readFileSync(this.statsFile, 'utf8'));
        } catch {
            return {};
        }
    }

    saveStats() {
        fs.writeFileSync(this.statsFile, JSON.stringify(this.stats, null, 2));
    }

    displayStats() {
        if (Object.keys(this.stats).length === 0) {
            console.log(colorize('Статистика пуста.', 'yellow'));
            return;
        }
        for (const [name, data] of Object.entries(this.stats)) {
            console.log(colorize(`${name}:`, 'bold'));
            console.log(`  Игр: ${data.games}, Побед: ${data.wins}, Поражений: ${data.losses}, Очки: ${data.score}`);
        }
    }

    loadChamber() {
        this.chamber = new Array(this.chambers).fill(0);
        const positions = Array.from({length: this.chambers}, (_,i) => i);
        for (let i = positions.length - 1; i > 0; i--) {
            const j = Math.floor(Math.random() * (i + 1));
            [positions[i], positions[j]] = [positions[j], positions[i]];
        }
        for (let i = 0; i < this.bullets; i++) {
            this.chamber[positions[i]] = 1;
        }
        this.position = 0;
    }

    spin() {
        this.position = Math.floor(Math.random() * this.chambers);
    }

    fire(player) {
        if (this.chamber[this.position] === 1) {
            this.history.push({player, shot: true});
            return true;
        } else {
            this.history.push({player, shot: false});
            this.position = (this.position + 1) % this.chambers;
            return false;
        }
    }

    updateStats(player, win, extra = 0) {
        const name = this.names[player];
        if (!this.stats[name]) {
            this.stats[name] = { games: 0, wins: 0, losses: 0, score: 0 };
        }
        this.stats[name].games++;
        if (win) this.stats[name].wins++;
        else this.stats[name].losses++;
        this.stats[name].score += extra;
        this.saveStats();
    }

    allPlayersShot() {
        if (this.history.length < this.playersCount) return false;
        const last = new Set();
        for (let i = this.history.length - this.playersCount; i < this.history.length; i++) {
            last.add(this.history[i].player);
        }
        return last.size === this.playersCount;
    }

    async playRound() {
        const rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });
        const question = (q) => new Promise(resolve => rl.question(q, resolve));

        if (this.mode === 'classic') {
            this.loadChamber();
            this.position = 0;
            while (true) {
                const player = this.current;
                console.log(colorize(`\nХод игрока ${this.names[player]}`, 'cyan'));
                await question('Нажмите Enter, чтобы выстрелить...');
                if (this.delay > 0) await new Promise(r => setTimeout(r, this.delay * 1000));
                const shot = this.fire(player);
                if (shot) {
                    console.log(colorize(`💥 БАХ! ${this.names[player]} проиграл!`, 'red'));
                    this.updateStats(player, false);
                    for (let i = 0; i < this.playersCount; i++) {
                        if (i !== player) this.updateStats(i, true);
                    }
                    rl.close();
                    return;
                } else {
                    console.log(colorize('Щелчок...', 'yellow'));
                    this.current = (this.current + 1) % this.playersCount;
                    if (this.allPlayersShot()) {
                        console.log(colorize('Все выжили! Ничья.', 'blue'));
                        for (let i = 0; i < this.playersCount; i++) {
                            this.updateStats(i, false);
                        }
                        rl.close();
                        return;
                    }
                }
            }
        } else if (this.mode === 'spin') {
            this.loadChamber();
            while (true) {
                const player = this.current;
                console.log(colorize(`\nХод игрока ${this.names[player]}`, 'cyan'));
                this.spin();
                await question('Нажмите Enter, чтобы выстрелить...');
                if (this.delay > 0) await new Promise(r => setTimeout(r, this.delay * 1000));
                const shot = this.fire(player);
                if (shot) {
                    console.log(colorize(`💥 БАХ! ${this.names[player]} проиграл!`, 'red'));
                    this.updateStats(player, false);
                    for (let i = 0; i < this.playersCount; i++) {
                        if (i !== player) this.updateStats(i, true);
                    }
                    rl.close();
                    return;
                } else {
                    console.log(colorize('Щелчок...', 'yellow'));
                    this.current = (this.current + 1) % this.playersCount;
                    if (this.allPlayersShot()) {
                        console.log(colorize('Все выжили! Ничья.', 'blue'));
                        for (let i = 0; i < this.playersCount; i++) {
                            this.updateStats(i, false);
                        }
                        rl.close();
                        return;
                    }
                }
            }
        } else if (this.mode === 'duel') {
            this.loadChamber();
            this.position = 0;
            while (true) {
                const player = this.current;
                console.log(colorize(`\nДуэль! Ход ${this.names[player]}`, 'cyan'));
                await question('Нажмите Enter, чтобы выстрелить...');
                if (this.delay > 0) await new Promise(r => setTimeout(r, this.delay * 1000));
                const shot = this.fire(player);
                if (shot) {
                    console.log(colorize(`💥 ${this.names[player]} проиграл дуэль!`, 'red'));
                    this.updateStats(player, false);
                    const other = 1 - player;
                    this.updateStats(other, true);
                    rl.close();
                    return;
                } else {
                    console.log(colorize('Щелчок...', 'yellow'));
                    this.current = (this.current + 1) % this.playersCount;
                }
            }
        } else if (this.mode === 'survival') {
            this.loadChamber();
            this.position = 0;
            const player = 0;
            let score = 0;
            console.log(colorize(`\nРежим выживания для ${this.names[player]}`, 'blue'));
            while (true) {
                console.log(colorize(`Текущий счёт: ${score}`, 'yellow'));
                const choice = await question('Выстрелить? (y/n): ');
                if (choice === 'q') {
                    console.log('Выход.');
                    rl.close();
                    return;
                }
                if (choice === 'n') {
                    console.log(colorize(`Вы остановились с ${score} очками.`, 'green'));
                    this.updateStats(player, true, score);
                    rl.close();
                    return;
                }
                if (choice === 'y') {
                    if (this.delay > 0) await new Promise(r => setTimeout(r, this.delay * 1000));
                    const shot = this.fire(player);
                    if (shot) {
                        console.log(colorize(`💥 БАХ! Вы проиграли! Очки: ${score}`, 'red'));
                        this.updateStats(player, false, score);
                        rl.close();
                        return;
                    } else {
                        console.log(colorize('Щелчок...', 'yellow'));
                        score++;
                        this.position = (this.position + 1) % this.chambers;
                    }
                }
            }
        }
        rl.close();
    }

    async play() {
        console.log(colorize('🔫 Добро пожаловать в русскую рулетку!', 'bold'));
        console.log(`Режим: ${this.mode}, Игроков: ${this.playersCount}, Патронов: ${this.bullets}, Гнёзд: ${this.chambers}`);
        if (this.delay > 0) console.log(`Задержка: ${this.delay*1000} мс`);
        console.log('Имена: ' + this.names.join(', '));
        console.log();
        await this.playRound();
    }
}

async function main() {
    const args = process.argv.slice(2);
    let mode = 'classic';
    let players = 2;
    let bullets = 1;
    let chambers = 6;
    let delay = 0.2;
    let names = null;
    let showStats = false;

    for (let i = 0; i < args.length; i++) {
        const arg = args[i];
        if (['classic', 'spin', 'duel', 'survival'].includes(arg)) {
            mode = arg;
        } else if (arg === '-p' && i+1 < args.length) {
            players = parseInt(args[++i]);
        } else if (arg === '-b' && i+1 < args.length) {
            bullets = parseInt(args[++i]);
        } else if (arg === '-c' && i+1 < args.length) {
            chambers = parseInt(args[++i]);
        } else if (arg === '-n' && i+1 < args.length) {
            names = args[++i].split(',');
        } else if (arg === '-d' && i+1 < args.length) {
            delay = parseFloat(args[++i]);
        } else if (arg === '-s') {
            showStats = true;
        } else if (arg === '-h') {
            console.log('Usage: node roulette.js [mode] [-p players] [-b bullets] [-c chambers] [-n names] [-d delay] [-s]');
            process.exit(0);
        }
    }
    if (!names) {
        names = Array.from({length: players}, (_,i) => `Игрок ${i+1}`);
    }
    const game = new Roulette(mode, players, bullets, chambers, delay, names);
    if (showStats) {
        game.displayStats();
        return;
    }
    await game.play();
}

main().catch(console.error);
